CMakeLists.txt

```diff
@@ -6,6 +6,8 @@ cmake_policy(SET CMP0057 NEW)
 
 project(opentelemetry-cpp)
 
+add_definitions(-DHAVE_CSTRING_TYPE)
```

Comment:
We were discussing the value of `CSTRING` with @pyohannes , and I think this is the best explanation I can come up with:
https://github.com/open-telemetry/opentelemetry-cpp/pull/519/files#r602669793

I do not have a very strong opinion on this, BUT if an outside container allows C-string type (for example, if original data structure is passed down either from C or from C# projection, or from any other language projection, it won't have `std::string` storage!!). Then copying it into `std::string` container implies a `memcpy`, that could have been otherwise avoided. I'd, again, refer to gRPC for example - which accepts strings as either `std::string` container or `const char *` pointer, both are strings from gRPC library perspective. I think we should respect similar contract in our library as well, as it gives us some future optimization opportunities where we can assume that possibly a `non-modifyable` for the duration of the `AddEvent` API call C-string has been passed to us. Any synchronous non-batching exporter may benefit from that by large, by avoiding the `memcpy` transform.

**All other exporters actually have been already modified to support it, I committed the relevant changes too.** Instead of removing this concept - we can remove the `#ifdef HAVE_CSTRING_TYPE` everywhere. I am proving by 100% successful CI in this PR here that the entire set of all existing tests, for all exporters, works great with this option defined.

Reply:
Ok, I'm fixing this.